src/coinsviewcacheasync.h

```diff
@@ -0,0 +1,241 @@
+     */
+    bool ProcessInputInBackground() const noexcept
+    {
+        const auto i{m_input_head.fetch_add(1, std::memory_order_relaxed)};
+        if (i >= m_inputs.size()) [[unlikely]] return false;
+
+        auto& input{m_inputs[i]};
+        // Inputs spending a coin from a tx earlier in the block won't be in the cache or db
+        if (std::ranges::binary_search(m_txids, input.outpoint.hash.ToUint256().GetUint64(0))) {
+            // We can use relaxed ordering here since we don't write the coin.
+            input.ready.test_and_set(std::memory_order_relaxed);
+            input.ready.notify_one();
+            return true;
+        }
+
+        if (auto coin{GetCoinWithoutMutating(input.outpoint)}) [[likely]] input.coin.emplace(std::move(*coin));
+        // We need release here, so writing coin in the line above happens before the main thread acquires.
+        input.ready.test_and_set(std::memory_order_release);
+        input.ready.notify_one();
+        return true;
+    }
+
+    //! Get the index in m_inputs for the given outpoint. Advances m_input_tail if found.
+    std::optional<uint32_t> GetInputIndex(const COutPoint &outpoint) const noexcept
+    {
+        for (const auto i : std::views::iota(m_input_tail, m_inputs.size())) {
+            if (m_inputs[i].outpoint == outpoint) {
+                m_input_tail = i + 1;
+                return i;
+            }
+        }
+        return std::nullopt;
+    }
+
+    CCoinsMap::iterator FetchCoin(const COutPoint &outpoint) const override
+    {
+        const auto [ret, inserted] = cacheCoins.try_emplace(outpoint);
+        if (!inserted) return ret;
+
+        if (const auto i{GetInputIndex(outpoint)}) [[likely]] {
+            auto& input{m_inputs[*i]};
+            // Check if the coin is ready to be read. We need to acquire to match the worker thread's release.
+            while (!input.ready.test(std::memory_order_acquire)) {
+                // Work instead of waiting if the coin is not ready
+                if (!ProcessInputInBackground()) {
+                    // No more work, just wait
+                    input.ready.wait(/*old=*/false, std::memory_order_acquire);
+                    break;
+                }
+            }
+            if (input.coin) [[likely]] ret->second.coin = std::move(*input.coin);
+        }
+
+        if (ret->second.coin.IsSpent()) [[unlikely]] {
+            // We will only get in here for BIP30 checks, txid collisions, or a block with missing or spent inputs.
+            if (auto coin{GetCoinWithoutMutating(outpoint)}) {
+                ret->second.coin = std::move(*coin);
+            } else {
+                cacheCoins.erase(ret);
+                return cacheCoins.end();
+            }
+        }
+
+        cachedCoinsUsage += ret->second.coin.DynamicMemoryUsage();
+        return ret;
+    }
+
+    std::vector<std::thread> m_worker_threads{};
+    std::barrier<> m_barrier;
```

Comment:
[we should](https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&branch=31132-69310ec0035351e486cfb51cb66492639f6f8b47&id=aureleoules_bitcoin&open=AZrV9y4tqc704xdUcnxD&tab=why) likely init it here instead:
```suggestion
    std::barrier<> m_barrier{WORKER_THREADS + 1};
```